File: gesture_to_flight/flight_instructions.py
#x-axis → forward/backward

#y-axis → left/ right

#z-axis → up/down

#roll → rotation about x-axis

#pitch → rotation about y-axis

#yaw → rotation about z-axis


import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from geometry_msgs.msg import PoseStamped
import tf_transformations
import math
import time


class FlightController(Node):
    def __init__(self):
        super().__init__('flight_controller')
        #whenever a message of type String is published to the topic /gesture_command, call the command_callback() function with the message
        self.sub = self.create_subscription(String, '/gesture_command', self.command_callback, 10)
        self.pos_sub = self.create_subscription(PoseStamped, '/mavros/local_position/pose', self.pose_callback, 10)

        self.pos_pub = self.create_publisher(PoseStamped, '/mavros/setpoint_position/local', 10)
        self.current_pose = PoseStamped()
        self.get_logger().info("FlightController ready and listening to gestures")

    def command_callback(self, msg):
        cmd = msg.data
        self.get_logger().info(f"Received command: {cmd}")
        if cmd == "PITCH_FORWARD":
            self.pmove_increment(10, 0, 0)
        elif cmd == "PITCH_BACKWARD":
            self.move_increment(-10, 0, 0)
        elif cmd == "ROLL_LEFT":
            self.bank(90, "LEFT")
        elif cmd == "ROLL_RIGHT":
            self.bank(90, "RIGHT")
        elif cmd == "THROTTLE_UP":
            self.move_increment(0, 0, 10)
        elif cmd == "THROTTLE_DOWN":
            self.move_increment(0, 0, -10)
        elif cmd == "YAW_LEFT":
            self.yaw(90, "LEFT")
        elif cmd == "YAW_RIGHT":
            self.yaw(90, "RIGHT")

        #need to finish hovering
        elif cmd == "HOLD_POSITION":
            self.get_logger().info("Holding position.")
        #need to finish kill
        elif cmd == "KILL":
            self.get_logger().warn("Emergency stop triggered!")

        #need to finish speeding up and down
        elif cmd == "SPEED_UP":
            pass
        
        elif cmd == "SPEED_DOWN":
            pass

    def bank(self, angle_deg, direction):
        angle_rad = math.radians(angle_deg if direction == "RIGHT" else -angle_deg)
        self.get_logger().info(f"Banking {direction} by {angle_deg} degrees")

        new_pose = PoseStamped()
        new_pose.header.stamp = self.get_clock().now().to_msg()
        new_pose.pose.position = self.current_pose.pose.position

        # Convert roll angle into quaternion (roll, pitch, yaw)
        q = tf_transformations.quaternion_from_euler(angle_rad, 0, 0)
        new_pose.pose.orientation.x = q[0]
        new_pose.pose.orientation.y = q[1]
        new_pose.pose.orientation.z = q[2]
        new_pose.pose.orientation.w = q[3]


    def yaw(self, angle_deg, direction):
        angle_rad = math.radians(angle_deg if direction == "RIGHT" else -angle_deg)
        self.get_logger().info(f"Yaw {direction} by {angle_deg} degrees")

        new_pose = PoseStamped()
        new_pose.header.stamp = self.get_clock().now().to_msg()
        new_pose.pose.position = self.current_pose.pose.position

        # Convert yaw angle into quaternion (roll, pitch, yaw)
        q = tf_transformations.quaternion_from_euler(0, 0, angle_rad)
        new_pose.pose.orientation.x = q[0]
        new_pose.pose.orientation.y = q[1]
        new_pose.pose.orientation.z = q[2]
        new_pose.pose.orientation.w = q[3]

    def move_increment(self, x, y, z):
        new_loc = PoseStamped()
        new_loc.header.stamp = self.get_clock().now().to_msg()
        new_loc.pose.position.x = self.current_pose.pose.position.x + x
        new_loc.pose.position.y = self.current_pose.pose.position.y + y
        new_loc.pose.position.z = self.current_pose.pose.position.z + z

        new_loc.pose.orientation = self.current_pose.pose.orientation
        self.pos_pub.publish(new_loc)
        self.get_logger().info(f"Moving drone to ({x}, {y}, {z})")

    def pose_callback(self, msg):
        self.current_pose = msg

def main(args=None):
    rclpy.init(args=args)
    node = FlightController()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: gesture_to_flight/src/flight_instructions.cpp
// C++ port of gesture_to_flight/flight_instructions.py
// Listens on /gesture_command (std_msgs/String) and /mavros/local_position/pose (PoseStamped)
// Publishes setpoints to /mavros/setpoint_position/local (PoseStamped)

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <cmath>
#include <algorithm>
#include <string>

class FlightController : public rclcpp::Node {
public:
  FlightController()
  : Node("flight_controller"), speed_factor_(1.0), min_speed_(0.1), max_speed_(3.0), pose_received_(false)
  {
    command_sub_ = this->create_subscription<std_msgs::msg::String>(
      "/gesture_command", 10,
      std::bind(&FlightController::command_callback, this, std::placeholders::_1));

    pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
      "/mavros/local_position/pose", 10,
      std::bind(&FlightController::pose_callback, this, std::placeholders::_1));

    pos_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>(
      "/mavros/setpoint_position/local", 10);

    RCLCPP_INFO(this->get_logger(), "FlightController ready and listening to gestures");
  }

private:
  void command_callback(const std_msgs::msg::String::SharedPtr msg)
  {
    std::string cmd = msg->data;
    RCLCPP_INFO(this->get_logger(), "Received command: %s", cmd.c_str());

    if (cmd == "PITCH_FORWARD") {
      move_increment(10.0 * speed_factor_, 0.0, 0.0);
    } else if (cmd == "PITCH_BACKWARD") {
      move_increment(-10.0 * speed_factor_, 0.0, 0.0);
    } else if (cmd == "ROLL_LEFT") {
      bank(90.0, "LEFT");
    } else if (cmd == "ROLL_RIGHT") {
      bank(90.0, "RIGHT");
    } else if (cmd == "THROTTLE_UP") {
      move_increment(0.0, 0.0, 10.0 * speed_factor_);
    } else if (cmd == "THROTTLE_DOWN") {
      move_increment(0.0, 0.0, -10.0 * speed_factor_);
    } else if (cmd == "YAW_LEFT") {
      yaw(90.0, "LEFT");
    } else if (cmd == "YAW_RIGHT") {
      yaw(90.0, "RIGHT");
    } else if (cmd == "HOLD_POSITION") {
      RCLCPP_INFO(this->get_logger(), "Holding position.");
      if (pose_received_) {
        pos_pub_->publish(current_pose_);
      } else {
        RCLCPP_WARN(this->get_logger(), "No current pose available yet, cannot hold position.");
      }
    } else if (cmd == "KILL") {
      RCLCPP_WARN(this->get_logger(), "Emergency stop triggered!");
      // Note: for a real emergency stop, additional actions should be taken (land/disarm)
    } else if (cmd == "SPEED_UP") {
      speed_factor_ = std::min(speed_factor_ + 0.1, max_speed_);
      RCLCPP_INFO(this->get_logger(), "Speed increased → %.2fx", speed_factor_);
    } else if (cmd == "SPEED_DOWN") {
      speed_factor_ = std::max(speed_factor_ - 0.1, min_speed_);
      RCLCPP_INFO(this->get_logger(), "Speed decreased → %.2fx", speed_factor_);
    } else {
      RCLCPP_WARN(this->get_logger(), "Unknown command: %s", cmd.c_str());
    }
  }

  void bank(double angle_deg, const std::string & direction)
  {
    double angle_rad = angle_deg * M_PI / 180.0;
    if (direction != "RIGHT") angle_rad = -angle_rad;
    RCLCPP_INFO(this->get_logger(), "Banking %s by %.1f degrees", direction.c_str(), angle_deg);

    if (!pose_received_) {
      RCLCPP_WARN(this->get_logger(), "No current pose available yet; cannot bank.");
      return;
    }

    geometry_msgs::msg::PoseStamped new_pose;
    new_pose.header.stamp = this->now().to_msg();
    new_pose.header.frame_id = current_pose_.header.frame_id;
    new_pose.pose.position = current_pose_.pose.position;

    tf2::Quaternion q;
    // roll, pitch, yaw
    q.setRPY(angle_rad, 0.0, 0.0);
    new_pose.pose.orientation.x = q.x();
    new_pose.pose.orientation.y = q.y();
    new_pose.pose.orientation.z = q.z();
    new_pose.pose.orientation.w = q.w();

    pos_pub_->publish(new_pose);
  }

  void yaw(double angle_deg, const std::string & direction)
  {
    double angle_rad = angle_deg * M_PI / 180.0;
    if (direction != "RIGHT") angle_rad = -angle_rad;
    RCLCPP_INFO(this->get_logger(), "Yaw %s by %.1f degrees", direction.c_str(), angle_deg);

    if (!pose_received_) {
      RCLCPP_WARN(this->get_logger(), "No current pose available yet; cannot yaw.");
      return;
    }

    geometry_msgs::msg::PoseStamped new_pose;
    new_pose.header.stamp = this->now().to_msg();
    new_pose.header.frame_id = current_pose_.header.frame_id;
    new_pose.pose.position = current_pose_.pose.position;

    tf2::Quaternion q;
    // roll, pitch, yaw
    q.setRPY(0.0, 0.0, angle_rad);
    new_pose.pose.orientation.x = q.x();
    new_pose.pose.orientation.y = q.y();
    new_pose.pose.orientation.z = q.z();
    new_pose.pose.orientation.w = q.w();

    pos_pub_->publish(new_pose);
  }

  void move_increment(double x, double y, double z)
  {
    if (!pose_received_) {
      RCLCPP_WARN(this->get_logger(), "No current pose available yet; cannot move.");
      return;
    }

    geometry_msgs::msg::PoseStamped new_loc;
    new_loc.header.stamp = this->now().to_msg();
    new_loc.header.frame_id = current_pose_.header.frame_id;
    new_loc.pose.position.x = current_pose_.pose.position.x + x;
    new_loc.pose.position.y = current_pose_.pose.position.y + y;
    new_loc.pose.position.z = current_pose_.pose.position.z + z;
    new_loc.pose.orientation = current_pose_.pose.orientation;

    pos_pub_->publish(new_loc);
    RCLCPP_INFO(this->get_logger(), "Moving drone by (%.2f, %.2f, %.2f)", x, y, z);
  }

  void pose_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
  {
    current_pose_ = *msg;
    pose_received_ = true;
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr command_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_sub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pos_pub_;

  geometry_msgs::msg::PoseStamped current_pose_;
  bool pose_received_;

  double speed_factor_;
  double min_speed_;
  double max_speed_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<FlightController>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
